### PR TITLE
`Communication`: Reduce number of websocket messages

### PR DESCRIPTION
**Pull request by lennart-keller**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

The server sends multiple WebSocket messages when creating new messages or replies to clients. This is a performance issue.
Additionally, a lot of post-related code is obsolete can be be removed. Tests can be adapted for messages.

### Description

For course-wide channels, the server sends one PostDTO to the course-wide topic `/topic/metis/courses/:courseId`. For any other non-course-wide conversation, the server sends one PostDTO to the user-specific topic `/topic/user/:userId/notifications/conversations`.
A PostDTO is of the form `public record PostDTO(Post post, MetisCrudAction action, Notification notification)`. 
Depending on the client state and on the user settings, the client can either display the new or updated post within a channel or display the provided notification instead.

The the obsolete server code for posts has been removed/adapter to only be compatible with plagiarism posts and conversation messages.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Tutor
- 1 Course

1. Write messages and replies on the messages tab in course-wide channels. All course members should receive a notification or directly see the message/reply if on the messages tab
2. Try out reactions to messages and replies
3. Try out deleting messages/replies
4. Try out pinning messages
      a. In channels, only moderators/instructors should be able to pin messages
      b. In group chats, only the creator should be able to pin messages
      c. In 1to1 chats, both members should be able to pin messages
5. Deactivate notifications for a user, they should not receive notifications anymore 

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [plagiarism-case-instructor-detail-view.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/plagiarism-cases/instructor-view/detail-view/plagiarism-case-instructor-detail-view.component.ts.html) | 98.87% | ✅ |
| [channel.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/entities/metis/conversation/channel.model.ts.html) | 100% | ✅ |
| [conversation-websocket-dto.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/entities/metis/conversation/conversation-websocket-dto.model.ts.html) | 100% | ✅ |
| [metis-post-dto.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/entities/metis/metis-post-dto.model.ts.html) | 0% | ✅ |
| [post.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/entities/metis/post.model.ts.html) | 100% | ✅ |
| [system-notification.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/entities/system-notification.model.ts.html) | 87.5% | ✅ |
| [conversation-messages.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/overview/course-conversations/layout/conversation-messages/conversation-messages.component.ts.html) | 94.05% | ✅ |
| [course-discussion.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/overview/course-discussion/course-discussion.component.ts.html) | 96.33% | ✅ |
| [discussion-section.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/overview/discussion-section/discussion-section.component.ts.html) | 93.75% | ✅ |
| [channel.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/conversations/channel.service.ts.html) | 100% | ✅ |
| [conversation.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/conversations/conversation.service.ts.html) | 96.55% | ✅ |
| [metis-conversation.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/metis-conversation.service.ts.html) | 84.21% | ✅ |
| [metis.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/metis.service.ts.html) | 91.05% | ✅ |
| [metis.util.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/metis.util.ts.html) | 100% | ✅ |
| [post.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/post.service.ts.html) | 80.39% | ✅ |
| [post.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/post/post.component.ts.html) | 92.15% | ✅ |
| [posting-content.components.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/posting-content/posting-content.components.ts.html) | 95.89% | ✅ |
| [post-create-edit-modal.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/posting-create-edit-modal/post-create-edit-modal/post-create-edit-modal.component.ts.html) | 92.53% | ✅ |
| [answer-post-header.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/posting-header/answer-post-header/answer-post-header.component.ts.html) | 100% | ✅ |
| [post-header.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/posting-header/post-header/post-header.component.ts.html) | 100% | ✅ |
| [post-reactions-bar.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/posting-reactions-bar/post-reactions-bar/post-reactions-bar.component.ts.html) | 87.93% | ✅ |
| [notification-popup.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/notification/notification-popup/notification-popup.component.ts.html) | 91.66% | ✅ |
| [notification.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/notification/notification.service.ts.html) | 80.42% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [User.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/User.html) | 92% | ✅ |
| [Post.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.metis/Post.html) | 97% | ✅ |
| [PostContextConstraintValidator.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.metis/PostContextConstraintValidator.html) | 100% | ✅ |
| [ConversationNotification.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.notification/ConversationNotification.html) | 89% | ✅ |
| [GroupNotificationFactory.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.notification/GroupNotificationFactory.html) | 94% | ✅ |
| [NotificationConstants.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.notification/NotificationConstants.html) | 96% | ✅ |
| [NotificationFactory.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.notification/NotificationFactory.html) | 100% | ✅ |
| [NotificationTargetFactory.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.notification/NotificationTargetFactory.html) | 99% | ✅ |
| [SingleUserNotificationFactory.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.notification/SingleUserNotificationFactory.html) | 89% | ✅ |
| [StatisticsRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/StatisticsRepository.html) | 98% | ✅ |
| [ConversationMessageRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository.metis/ConversationMessageRepository.html) | 93% | ✅ |
| [PostRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository.metis/PostRepository.html) | 81% | ✅ |
| [ConversationRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository.metis.conversation/ConversationRepository.html) | 56% | ✅ |
| [PostSpecs.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository.specs/PostSpecs.html) | 87% | ✅ |
| [AnswerMessageService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.metis/AnswerMessageService.html) | 88% | ✅ |
| [AnswerPostService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.metis/AnswerPostService.html) | 94% | ✅ |
| [ConversationMessagingService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.metis/ConversationMessagingService.html) | 93% | ✅ |
| [PostService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.metis/PostService.html) | 87% | ✅ |
| [PostingService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.metis/PostingService.html) | 99% | ✅ |
| [ReactionService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.metis/ReactionService.html) | 100% | ✅ |
| [ChannelService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.metis.conversation/ChannelService.html) | 94% | ✅ |
| [ConversationNotificationService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.notifications/ConversationNotificationService.html) | 97% | ✅ |
| [GroupNotificationService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.notifications/GroupNotificationService.html) | 98% | ✅ |
| [MailService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.notifications/MailService.html) | 95% | ✅ |
| [SingleUserNotificationService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.notifications/SingleUserNotificationService.html) | 96% | ✅ |
| [PostContextFilter.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.dto/PostContextFilter.html) | 95% | ✅ |
| [ConversationMessageResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.metis/ConversationMessageResource.html) | 91% | ✅ |
| [PostResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.metis/PostResource.html) | 100% | ✅ |
| [ChannelResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.metis.conversation/ChannelResource.html) | 92% | ✅ |
| [ConversationWebsocketDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.websocket.dto.metis/ConversationWebsocketDTO.html) | 100% | ✅ |
| [PostDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5876/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.websocket.dto.metis/PostDTO.html) | 100% | ✅ |

### Screenshots

No UI changes

### CODE REVIEW

**Review by rriyaldhi — COMMENTED** [review]
Only found 1 error during testing. Everything looks good.

image

**Comment by lennart-keller** [comment]
> Only found 1 error during testing. Everything looks good.
> 
> image

Nice catch, thank you. I have fixed this issue :)

**Review by milljoniaer — APPROVED** [review]
Code LGTM 🚀

**Review by rriyaldhi — DISMISSED** [review]
Code looks good 👍

**Review by basak-akan — APPROVED** [review]
Tested on ts3 and worked well for below scenarios

- Course-wide messaging, everyone notified or sees messages/replies.
- Reactions, worked well
- Message deletion worked well.
- Pinning messages also worked as described as rules per chat type (channels, group, 1-to-1)
- Also played around with notification settings such as user mentions and conversation settings, also worked well

File: src/main/webapp/app/shared/metis/post/post.component.html [hunk 1]
```diff
@@ -52,7 +52,7 @@
                     @if (!isCourseMessagesPage && isCommunicationPage) {
                         <a
                             class="reference-hash"
-                            [routerLink]="metisService.getLinkForPost(posting)"
+                            [routerLink]="metisService.getLinkForPost()"
```
Comment (milljoniaer) [inline src/main/webapp/app/shared/metis/post/post.component.html:55]: can we get rid of this function call?

**Review by milljoniaer — APPROVED** [review]
Code LGTM, just left one comment

**Review by basak-akan — APPROVED** [review]
Re-tested on TS5 with scenarios in PR description, all worked well for communication enabled/disabled and messaging enabled/disabled settings.

**Review by rriyaldhi — APPROVED** [review]
Reapprove code

**Review by Ceiridge — APPROVED** [review]
Tested on ts5. It works as expected, but I encountered a problem searching for users when creating a group chat.
Searching for `Test` or `User` or the user number didn't work, but searching for `Artemis` gave me a user list. Maybe I should create an issue on this or is this related to this PR feature?

image
image

**Comment by lennart-keller** [comment]
> Tested on ts5. It works as expected, but I encountered a problem searching for users when creating a group chat. Searching for `Test` or `User` or the user number didn't work, but searching for `Artemis` gave me a user list. Maybe I should create an issue on this or is this related to this PR feature?
> 
> image image

Thank you for your approve. The issue you described is unrelated to this PR and also appears when searching for users you want to add to a course. However, when deploying this PR to TS2, it works with `Test`, `User`, etc. for both scenarios. I guess it has something to do with MySQL/Postgres and the way the search queries are written (I haven't edited them in this PR). Would be nice if you could create an issue for this :)

**Review by rstief — APPROVED** [review]
Re-tested on ts2 and all previous issues seem to be fixed. Looks good!

File: src/main/webapp/app/shared/metis/post/post.component.html [hunk 2]
```diff
@@ -50,12 +50,7 @@
                     }
                     <!-- show id only on communication page -->
                     @if (!isCourseMessagesPage && isCommunicationPage) {
-                        <a
-                            class="reference-hash"
-                            [routerLink]="metisService.getLinkForPost()"
```
Comment (milljoniaer) [inline src/main/webapp/app/shared/metis/post/post.component.html]: nice, thanks for removing this

**Review by milljoniaer — COMMENTED** [review]
Changes look good, thanks for implementing these 🚀

**Review by milljoniaer — APPROVED** [review]

**Review by Kroko-fant — APPROVED** [review]
tested on ts4 in a testing session worked without any issues

**Review by kaancayli — APPROVED** [review]
tested on ts4 in a testing session worked without any issues for me as well.

**Review by yassinsws — APPROVED** [review]
tested in a testing session works

**Comment by nata-81** [comment]
tested on ts 4. Works expected.

**Review by nata-81 — APPROVED** [review]
tested on ts 4. Works expected.

**Review by MaximilianJG — COMMENTED** [review]
Manually tested on Ts2. Works as expected

File: src/main/java/de/tum/in/www1/artemis/web/websocket/dto/metis/PostDTO.java [hunk 3]
```diff
@@ -3,11 +3,15 @@
...
 
 import de.tum.in.www1.artemis.domain.metis.Post;
+import de.tum.in.www1.artemis.domain.notification.Notification;
 
 /**
  * DTO that is included as payload for post related websocket messages
  */
 @JsonInclude(JsonInclude.Include.NON_EMPTY)
-public record PostDTO(Post post, MetisCrudAction action) {
+public record PostDTO(Post post, MetisCrudAction action, Notification notification) {
```
Comment (nityanandaz) [inline src/main/java/de/tum/in/www1/artemis/web/websocket/dto/metis/PostDTO.java:12]: New DTO field

File: src/main/webapp/app/entities/metis/conversation/channel.model.ts [hunk 4]
```diff
@@ -22,6 +22,7 @@ export class Channel extends Conversation {
     public isPublic?: boolean;
     public isAnnouncementChannel?: boolean;
     public isArchived?: boolean;
+    public isCourseWide?: boolean;
```
Comment (nityanandaz) [inline src/main/webapp/app/entities/metis/conversation/channel.model.ts:25]: New channel field

File: src/main/webapp/app/overview/course-conversations/layout/conversation-selection-sidebar/conversation-sidebar-section/conversation-sidebar-entry/conversation-sidebar-entry.component.html [hunk 5]
```diff
@@ -4,9 +4,9 @@
             <span class="conversation-name">
                 @if (conversationAsChannel) {
                     <jhi-channel-icon
-                        [isPublic]="conversationAsChannel.isPublic!"
-                        [isArchived]="conversationAsChannel.isArchived!"
-                        [isAnnouncementChannel]="conversationAsChannel.isAnnouncementChannel!"
+                        [isPublic]="(conversationAsChannel.isPublic || conversationAsChannel.isCourseWide) ?? false"
```
Comment (nityanandaz) [inline src/main/webapp/app/overview/course-conversations/layout/conversation-selection-sidebar/conversation-sidebar-section/conversation-sidebar-entry/conversation-sidebar-entry.component.html:7]: New public condition

File: src/test/javascript/spec/component/overview/course-conversations/services/conversation.service.spec.ts [hunk 6]
```diff
@@ -19,6 +21,7 @@ describe('ConversationService', () => {
             providers: [
                 { provide: TranslateService, useClass: MockTranslateService },
                 { provide: AccountService, useClass: MockAccountService },
+                { provide: NotificationService, useClass: MockNotificationService },
```
Comment (nityanandaz) [inline src/test/javascript/spec/component/overview/course-conversations/services/conversation.service.spec.ts:24]: `{ provide: NotificationService, useClass: MockNotificationService }` or `MockProvider(NotificationService)`?

#\7535

**Review by nityanandaz — COMMENTED** [review]
Notes to myself, native client POIs:

**Pull request merged**